embassy/src/executor/arch/xtensa.rs

```diff
@@ -0,0 +1,72 @@
+use core::marker::PhantomData;
+use core::ptr;
+
+use atomic_polyfill::{AtomicBool, Ordering};
+
+use super::{raw, Spawner};
+
+/// global atomic used to keep track of whether there is work to do since sev() is not available on Xtensa
+///
+static SIGNAL_WORK_THREAD_MODE: AtomicBool = AtomicBool::new(false);
+
+/// Xtensa Executor
+pub struct Executor {
+    inner: raw::Executor,
+    not_send: PhantomData<*mut ()>,
+}
+
+impl Executor {
+    /// Create a new Executor.
+    pub fn new() -> Self {
+        Self {
+            // use Signal_Work_Thread_Mode as substitute for local interrupt register
+            inner: raw::Executor::new(
+                |_| {
+                    SIGNAL_WORK_THREAD_MODE.store(true, Ordering::SeqCst);
+                },
+                ptr::null_mut(),
+            ),
+            not_send: PhantomData,
+        }
+    }
+
+    /// Run the executor.
+    ///
+    /// The `init` closure is called with a [`Spawner`] that spawns tasks on
+    /// this executor. Use it to spawn the initial task(s). After `init` returns,
+    /// the executor starts running the tasks.
+    ///
+    /// To spawn more tasks later, you may keep copies of the [`Spawner`] (it is `Copy`),
+    /// for example by passing it as an argument to the initial tasks.
+    ///
+    /// This function requires `&'static mut self`. This means you have to store the
+    /// Executor instance in a place where it'll live forever and grants you mutable
+    /// access. There's a few ways to do this:
+    ///
+    /// - a [Forever](crate::util::Forever) (safe)
+    /// - a `static mut` (unsafe)
+    /// - a local variable in a function you know never returns (like `fn main() -> !`), upgrading its lifetime with `transmute`. (unsafe)
+    ///
+    /// This function never returns.
+    pub fn run(&'static mut self, init: impl FnOnce(Spawner)) -> ! {
+        init(self.inner.spawner());
+
+        loop {
+            unsafe {
+                self.inner.poll();
+                // we do not care about race conditions between the load and store operations, interrupts
+                // will only set this value to true.
+                // if there is work to do, loop back to polling
+                // TODO can we relax this?
+                if SIGNAL_WORK_THREAD_MODE.load(Ordering::SeqCst) {
+                    SIGNAL_WORK_THREAD_MODE.store(false, Ordering::SeqCst);
+                }
+                // if not, wait for interrupt
+                else {
+                    // waiti sets the PS.INTLEVEL therefore no critical section is needed
```

Comment:
So in my testing, I've verified that the TRM is indeed correct. However, I also found some more notes on implemented critical sections on Xtensa, they should actually be done by increasing the `PS.INTLEVEL` such that `PS.INTLEVEL > any possible interrupt level` instead of disabling interrupts via the `INTENABLE` register. I need to do a bit more reading to verify that no interrupts can exist at the max (which is 15), the TRM is kinda vague on that. I guess NMI might be at that level, but what can ya do about that :D. By modifying `PS.INTLEVEL`, the `waiti` instruction can atomically reduce the `PS.INTLEVEL` whilst slipping into sleep mode, hence no race conditions :D.

So what this actually means is we need `critical_section` impls for Xtensa. It seems the path forward for `critical_section` is that every HAL/PAC will have a per chip cs impl that can handle multi core etc, so I've opened https://github.com/esp-rs/esp-hal/issues/87.

I'll make sure all this theory work actually works on hardware and get back to you :).